### PR TITLE
BACKEND-7509 :: Feature :: Harmony PHP | PHP 8.1

### PR DESCRIPTION
**Pull request by Alex-DA**
Main Asana Task: https://app.asana.com/0/1109863238977521/1203478746427509/f
Related tasks:
- https://app.asana.com/0/1109863238977521/1202936325763966/f
- https://app.asana.com/0/1109863238977521/1202798243156793/f
- https://app.asana.com/0/1109863238977521/1203446564324737/f

### Fixes

1. Fixed PutAll in MySQL DataSource and InMemory DataSource
2. To avoid inconsistencies in data, added as requirement `OrderBy` on Queries with `offset` and `limit`
3. Renamed confusing variables in `core/src/Repository/DataSource/Sql/Helper/SqlBuilder.php`
4. Renamed confusing variables in `core/src/Repository/RepositoryMapper.php`
5. Fixed some docker issues with Sample project

### Features

1. Now on MySQL DataSource we can use a custom class instead of `object`. Ex. `sample/src/app/Product/Data/Sql/Mapper/ProductEntityToSqlDataMapper.php`. This allow us to expect always a fixed data from SQL Queries, before we need to double check if a column/attribute exist on a `object`.
2. Refactor Sample project to align with PHP 8.1 features. Ex. `sample/src/app/Product/Data/Entity/ProductEntity.php`
3. Refactor Core code to align with PHP 8.1 features. Ex. Changed all immutable attributes to `readonly`
4. Upgrade Docker to PHP 8.1
5. Static Analysis and Prettier to PHP 8.1
6. Added Code Style GitHub Action with Prettier PHP
8. Added Code Style hook in commits with Prettier PHP
9. Applied Prettier to all files
10. Changed Static Analysis of PHPStan and PSalm to Maximum level
11. Code Coverage from 25% to 47%
12. New Integration Tests using a MySQL DB. Modified docker-compose, new test-integration.sh script, new Schema example for Sample project, etc
13. Now we can test DataSource extending `sample/src/tests/unit/Product/Domain/Interactor/ProductInteractorsTest.php` -> `sample/src/tests/unit/Product/Domain/Interactor/InMemoryProductInteractorsTest.php` and `sample/src/tests/integration/Product/Domain/Interactor/RawSqlProductInteractorsTest.php`
14. New Helper for testing `core/src/Helper/Random.php + Format.php`
15. Added PHPStorm default config for: testing, run project, code style
16. Upgrade NVMRC to 18.10

### CODE REVIEW

**Comment by joselufo** [comment]
@Alex-DA I converted it to Draft as the title said WIP.

File: sample/src/tests/integration/Product/Domain/Interactor/RawSqlProductInteractorsTest.php [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+
+  /**
+   * @psalm-suppress InvalidArgument
+   */
+  protected function getProvider(): ProductProvider {
+    $sqlBuilder = new SqlBuilder(new ProductSqlSchema(), new QueryFactory());
+    $pdoWrapper = new PdoWrapper($this->getPdo());
+    $dataSource = new RawSqlDataSource($pdoWrapper, $sqlBuilder);
+
+    $dataSourceMapper = new DataSourceMapper(
```
Comment (urijp) [inline sample/src/tests/integration/Product/Domain/Interactor/RawSqlProductInteractorsTest.php:31]: Same as before, like you call productProvider some lines below, here is $productDatasource

File: sample/src/tests/unit/Product/Domain/Interactor/ProductInteractorsTest.php [hunk 2]
```diff
@@ -0,0 +1,160 @@
...
+use Sample\Product\Domain\Model\Product;
+use Sample\Product\ProductProvider;
+
+abstract class ProductInteractorsTest extends TestCase {
+  abstract protected function getProvider(): ProductProvider;
+
+  public function testCreateProductInteractor(): void {
+    $productProvider = $this->getProvider();
+    $product = $this->getProductOne();
+    $productSaved = $this->putProduct($productProvider, $product);
```
Comment (urijp) [inline sample/src/tests/unit/Product/Domain/Interactor/ProductInteractorsTest.php:19]: These three lines are repeated a lot, could we move them to a new method?

File: sample/src/tests/unit/Product/Domain/Interactor/ProductInteractorsTest.php [hunk 3]
```diff
@@ -0,0 +1,160 @@
...
+
+    $productWithChange = new Product(
+      id: $productSaved->id,
+      name: Random::generateAlphaNumeric(),
+      description: Random::generateAlphaNumeric(),
+      price: $productSaved->price,
+    );
+
+    $productUpdated = $this->putProduct($productProvider, $productWithChange);
```
Comment (urijp) [inline sample/src/tests/unit/Product/Domain/Interactor/ProductInteractorsTest.php:47]: updatedProduct

File: sample/src/tests/unit/Product/Domain/Interactor/ProductInteractorsTest.php [hunk 4]
```diff
@@ -0,0 +1,160 @@
...
+
+    $this->assertEquals($product, $productGot);
+  }
+
+  public function testUpdateProductInteractor(): void {
+    $productProvider = $this->getProvider();
+    $product = $this->getProductOne();
+    $productSaved = $this->putProduct($productProvider, $product);
+
+    $productWithChange = new Product(
```
Comment (urijp) [inline sample/src/tests/unit/Product/Domain/Interactor/ProductInteractorsTest.php:40]: editedProduct

File: sample/src/tests/unit/Product/Domain/Interactor/ProductInteractorsTest.php [hunk 5]
```diff
@@ -0,0 +1,160 @@
...
+      description: Random::generateAlphaNumeric(),
+      price: $productSaved->price,
+    );
+
+    $productUpdated = $this->putProduct($productProvider, $productWithChange);
+
+    $this->assertEquals($productWithChange, $productUpdated);
+
+    $queryGet = new KeyQuery((string) $productSaved->id);
+    $productGot = $productProvider->provideGetInteractor()($queryGet);
```
Comment (urijp) [inline sample/src/tests/unit/Product/Domain/Interactor/ProductInteractorsTest.php:52]: $resultingProduct maybe?

File: sample/src/tests/unit/Product/Domain/Interactor/ProductInteractorsTest.php [hunk 6]
```diff
@@ -0,0 +1,160 @@
...
+
+    $this->assertEquals($productWithChange, $productGot);
+  }
+
+  public function testCreateAllProductsInteractor(): void {
+    $productProvider = $this->getProvider();
+    $products = $this->getListOfProducts();
+    $productsSaved = $this->putProducts($productProvider, $products);
```
Comment (urijp) [inline sample/src/tests/unit/Product/Domain/Interactor/ProductInteractorsTest.php:60]: $savedProducts

**Review by urijp — CHANGES_REQUESTED** [review]

File: core/src/Domain/Interactor/PutAllInteractor.php [hunk 7]
```diff
@@ -15,20 +15,20 @@ class PutAllInteractor {
...
    */
-  public function __construct(protected PutRepository $putRepository) {
+  public function __construct(protected readonly PutRepository $putRepository) {
   }
 
   /**
-   * @param array<T>  $models
-   * @param Query|null $query
+   * @param array<T>       $models
+   * @param Query|null     $query
```
[the remark this answers is not in the document]
Comment (Alex-DA) [reply core/src/Domain/Interactor/PutAllInteractor.php:23]: @urijp prettier.

File: sample/src/app/Application/ProductIndexAction.php [hunk 8]
```diff
@@ -0,0 +1,127 @@
...
+    return $product;
+  }
+
+  /**
+   * @param Product[] $products
+   *
+   * @return Product[]
+   */
+  protected function putAllProductsAction(array $products): array {
+    $query = new AllQuery();
```
[the remark this answers is not in the document]
Comment (Alex-DA) [reply sample/src/app/Application/ProductIndexAction.php:82]: @orioljp the default one is VoidQuery in `core/src/Domain/Interactor/PutAllInteractor.php`

File: sample/src/app/Application/ProductIndexAction.php [hunk 9]
```diff
@@ -0,0 +1,127 @@
...
+    $result = ($this->putAllProduct)($products, $query);
+
+    return $result;
+  }
+
+  /**
+   * @return Product[]
+   */
+  protected function getAllProductsAction(): array {
+    $query = new AllQuery();
```
[the remark this answers is not in the document]
Comment (Alex-DA) [reply sample/src/app/Application/ProductIndexAction.php:92]: Thanks for your comment, I found a mismatch on GetAllInteractor, I changed the default query from AllQuery to VoidQuery.

File: sample/src/app/Application/ProductIndexAction.php (hunk 8 above)
[the remark this answers is not in the document]
Comment (urijp) [reply sample/src/app/Application/ProductIndexAction.php:82]: But what's the default behaviour? Here we want to put all with a put all method, that should be the default behaviour so no query should be needed.

**Review by urijp — APPROVED** [review]

File: sample/src/app/Application/ProductIndexAction.php (hunk 8 above)
[the remark this answers is not in the document]
Comment (Alex-DA) [reply sample/src/app/Application/ProductIndexAction.php:82]: @urijp I will create a new task to follow-up this case, as I see that we have many strategies in the other Harmony libs, and I prefer to align all the same.

File: core/src/Helper/Random.php [hunk 10]
```diff
@@ -0,0 +1,35 @@
...
+/**
+ * @see RandomTest
+ */
+class Random {
+  /**
+   * @link https://symfony.com/doc/5.2/components/security/secure_tools.html#generating-a-secure-random-number
+   */
+  public static function generateInteger(): int {
+    $randomInteger = random_int(1, mt_getrandmax());
+
```
[the remark this answers is not in the document]
Comment (Alex-DA) [reply core/src/Helper/Random.php:15]: @magarzon Is not a Harmony Code Style, but my personal Code Style 😝
The reason is to make work easier when we are debugging with XDebug.

Normally when you are debugging a result from a function you need to modify the code first to get the result in a variable, add a breakpoint and see the value, and finally revert the changes.

Using always an intermediate variable before returns, you only need to add a breakpoint to the `return` line and continue working.

I know that in other languages the debugger allows you to stop in the last line and see the result before the return, but is not the case with XDebug yet (at least, last time that I checked).

**Pull request merged**